Introduce fastpath backend

kurtmcmillan opened the pull request:
## Motivation

Avoid bouncing IO through host memory when the destination buffer is on a GPU.

kurtmcmillan committed:
rocfile: Introduce fastpath backend

derobins commented on rocfile/src/backend/fastpath.cpp:
```diff
@@ -0,0 +1,142 @@
+ *      - size_copied and status are never modified
+ *      - returns false (error) if hsa does not return HSA_STATUS_SUCCESS
+ *  - hsa (userspace)
+ *      - repository: rocm-systems
```
Probably need to add hsa to the codespell dotfile

kurtmcmillan replied:
Added.

kurtmcmillan committed:
rocfile: Update fastpath notes

kurtmcmillan committed:
codespell: Add HSA to the ignore list

kurtmcmillan committed:
review: IWYU fixes after rebase

jordan-turbofish commented on rocfile/src/backend/fastpath.cpp:
```diff
@@ -0,0 +1,142 @@
+using namespace std;
+
+/* The fastpath backend is used when:
+ *  - IO is alligned (size, file_offset, buffer_offset)
```
alligned -> aligned

derobins commented on rocfile/src/backend/fastpath.cpp:
```diff
@@ -0,0 +1,157 @@
+#include <fcntl.h>
+#include <hip/hip_runtime_api.h>
+#include <stdexcept>
+#include <stdint.h>
```
Should be `<cstdint>`

kurtmcmillan replied:
Fixed.

kurtmcmillan committed:
review: Fix backend initialization

Moving backend initialization out of the DriverState constructor and into
DriverState::incRefCount() resulted in the backends getting initialized after
rocFileIo() cached an empty backend vector from DriverState::getBackends().

This change relies on compilers supporting thread-stafe static initialization.
DriverState's backend vector is now initialized the first time
DriverState::getBackends() is called.

derobins commented on rocfile/src/backend/fastpath.cpp:
```diff
@@ -0,0 +1,157 @@
+
+int
+Fastpath::score(shared_ptr<IFile> file, shared_ptr<IBuffer> buffer, size_t size, off_t file_offset,
+                off_t buffer_offset) const
```
Should this be `hoff_t` ? (throughout the PR)

I only kept `off_t` in the system call equivalents since pread, mmap, etc. do not exist on Windows.

kurtmcmillan replied:
Changed to use hoff_t.

derobins commented on rocfile/src/backend/fastpath.cpp:
```diff
@@ -0,0 +1,157 @@
+/* The fastpath backend is used when:
+ *  - The file has been opened with the O_DIRECT flag
+ *  - IO is 4K aligned (size, file_offset, buffer_offset)
+ *      - This requirement will be relaxed in the future
```
It's probably more accurate to say that the buffer must be 4 kiB aligned and a multiple of 4 kiB

jordan-turbofish replied:
When you say multiple of 4KiB, are you talking about the length of the buffer?

kurtmcmillan replied:
Updated to read:

```
/* The fastpath backend is used when:
 *  - The file has been opened with the O_DIRECT flag
 *  - file_offset is 4KiB aligned
 *  - buffer_offset is 4KiB aligned
 *  - size if a multiple of 4KiB
 *  - The buffer type is hipMemoryTypeDevice
```

kurtmcmillan committed:
hipfile/examples/aiscp: Improve error messages

derobins commented on rocfile/src/backend/fastpath.h:
```diff
@@ -0,0 +1,35 @@
+#include "backend.h"
+
+#include <sys/types.h>
+#include <memory>
```
Not in alphabetical order

kurtmcmillan replied:
Fixed.

derobins commented on rocfile/test/fastpath.cpp:
```diff
@@ -0,0 +1,354 @@
+}
+
+// Ensure hip errors thrown by Hip::hipAmdFileRead/Hip::hipAmdFileWtite are not masked
+TEST_P(FastpathIoParam, IoDoesNotMaskHipRuntimeError)
```
typo: hipAmdFileWrite

kurtmcmillan replied:
Fixed

derobins commented on rocfile/test/fastpath.cpp:
```diff
@@ -0,0 +1,354 @@
+}
+
+// Ensure hip errors thrown by Hip::hipAmdFileRead/Hip::hipAmdFileWtite are not masked
+TEST_P(FastpathIoParam, IoDoesNotMaskSystemError)
```
Same typo here

kurtmcmillan replied:
Fixed

derobins commented on rocfile/test/fastpath.cpp:
```diff
@@ -0,0 +1,354 @@
+#include "rocfile-test.h"
+
+#include <array>
+#include <errno.h>
```
Should be `<cerrno>`

kurtmcmillan replied:
Fixed.

kurtmcmillan committed:
review: Fix broken unit tests

kurtmcmillan committed:
review: Address review comments

kurtmcmillan committed:
review: IWYU fix

derobins commented on rocfile/src/backend/fastpath.cpp:
```diff
@@ -0,0 +1,158 @@
+ *  - The file has been opened with the O_DIRECT flag
+ *  - file_offset is 4KiB aligned
+ *  - buffer_offset is 4KiB aligned
+ *  - size if a multiple of 4KiB
```
if --> is

kurtmcmillan replied:
Fixed.

kurtmcmillan committed:
review: Fix if --> is

kurtmcmillan committed:
Merge branch 'develop' into z/kumcmill/fastpath

kurtmcmillan committed:
Merge branch 'develop' into z/kumcmill/fastpath
